homeassistant/components/iotawatt/sensor.py

```diff
@@ -206,13 +228,39 @@ def extra_state_attributes(self):
         attrs = {"type": data.getType()}
         if attrs["type"] == "Input":
             attrs["channel"] = data.getChannel()
+        if self._accumulating:
+            attrs[
+                ATTR_LAST_UPDATE
+            ] = self.coordinator.api.getLastUpdateTime().isoformat()
 
         return attrs
 
+    async def async_added_to_hass(self):
+        """Load the last known state value of the entity if the accumulated type."""
+        await super().async_added_to_hass()
+        if self._accumulating:
+            state = await self.async_get_last_state()
+            self._accumulated_value = 0.0
+            if state:
+                try:
+                    self._accumulated_value = float(state.state)
+                    if ATTR_LAST_UPDATE in state.attributes:
+                        self.coordinator.update_last_run(
+                            dt.parse_datetime(state.attributes.get(ATTR_LAST_UPDATE))
+                        )
```

Comment:
If I'm reading this right, it looks like every entity that is restored is going to call `self.coordinator.update_last_run` which means its going to get called multiple times?

Reply:
Not quite.
coordinator.update_last_run only let the coordinator knows that the next run will need to retrieve data from that time. While each sensor should have exactly the same last_run value; it doesn't have to be (either inconsistency in the data, a sensor disabled and re-enable later etc).

So the coordinator will restore from the most recent sensor and update all the sensors from that point in time.

I'm not quite sure how I could get the coordinator to store that time and restore it. 

The refresh only occurs once.

Reply:
Ideally the `iotawattpy` package would take care of remembering and restoring these values which would avoid the need to have much of this logic in the integration itself.

You could trigger it to write it the state to disk on the `EVENT_HOMEASSISTANT_FINAL_WRITE` event.

The disadvantage to this approach vs what you have now is that `helpers.restore_state` writes to disk every 15 minutes which means less data is at risk of being lost if home assistant doesn't shutdown cleanly.

Reply:
The iotawattpy is a stateless library; it's only designed to retrieve the current data and that's it.
It has no room for having persistent state , and I don't think you would want it to.

Works just like a fancy WGET that also happen to parse the output data ready for HA to use it. It's much more logical for HA to do it, it has all the infrastructure to do so.

In a sense, this works in a similar fashion to the `integration` Riemann integration, it stores the time of the last update and when it gets a new one calculate the sum.